src/main/java/org/elasticsearch/action/fieldstats/FieldStats.java

```diff
@@ -0,0 +1,452 @@
+    private long sumTotalTermFreq;
+    protected T minValue;
+    protected T maxValue;
+
+    protected FieldStats() {
+    }
+
+    protected FieldStats(int type, long maxDoc, long docCount, long sumDocFreq, long sumTotalTermFreq) {
+        this.type = (byte) type;
+        this.maxDoc = maxDoc;
+        this.docCount = docCount;
+        this.sumDocFreq = sumDocFreq;
+        this.sumTotalTermFreq = sumTotalTermFreq;
+    }
+
+    /**
+     * @return the total number of documents.
+     *
+     * Note that, documents marked as deleted that haven't yet been merged way aren't taken into account.
+     */
+    public long getMaxDoc() {
+        return maxDoc;
+    }
+
+    /**
+     * @return the number of documents that have at least one term for this field, or -1 if this measurement isn't available.
+     *
+     * Note that, documents marked as deleted that haven't yet been merged way aren't taken into account.
+     */
+    public long getDocCount() {
+        return docCount;
+    }
+
+    /**
+     * @return A ratio as percentage of between all the documents that have at least one value for this field and all documents.
+     * Useful to determine whether a field is sparse or dense. (doc_count / max_doc)
+     */
+    public int getFieldRatio() {
+        if (docCount < 0 || maxDoc <= 0) {
+            return -1;
+        }
+        return (int) (docCount * 100 / maxDoc);
+    }
+
+    /**
+     * @return the sum of each term's document frequency in this field, or -1 if this measurement isn't available.
+     * Document frequency is the number of documents containing a particular term.
+     *
+     * Note that, documents marked as deleted that haven't yet been merged way aren't taken into account.
+     */
+    public long getSumDocFreq() {
+        return sumDocFreq;
+    }
+
+    /**
+     * @return the sum of the term frequencies of all terms in this field across all documents, or -1 if this measurement
+     * isn't available. Term frequency is the total number of occurrences of a term in a particular document and field.
+     *
+     * Note that, documents marked as deleted that haven't yet been merged way aren't taken into account.
+     */
+    public long getSumTotalTermFreq() {
+        return sumTotalTermFreq;
+    }
+
+    /**
+     * @return the lowest value in the field represented as a string.
+     *
+     * Note that, documents marked as deleted that haven't yet been merged way aren't taken into account.
+     */
+    public abstract String getMinValue();
+
+    /**
+     * @return the highest value in the field represented as a string.
+     *
+     * Note that, documents marked as deleted that haven't yet been merged way aren't taken into account.
+     */
+    public abstract String getMaxValue();
+
+    /**
+     * Merges the provided stats into this stats instance.
+     */
+    public void append(FieldStats stats) {
+        this.maxDoc += stats.maxDoc;
+        if (stats.docCount != -1) {
```

Comment:
maybe it is a matter of style, but i think its easier to handle the exceptional case like a guard up front: check stats.docCount == -1 and set to -1, otherwise sum. this is not really important to me.